README.md
# Factorio Mod: BetterBots
Adds useful new technologies to the tech tree that improve the existing bot experience. Updated to 1.1.x by [jollywalrus](https://mods.factorio.com/mod/BetterBots/discussion/5fcbab5123a3f47151037856) and Walkman100

This is [BetterBots](https://mods.factorio.com/mod/BetterBots) Updated to 1.1.x


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.3.11
Date: 2021.09.05
  Changes:
    - Added Wired Roboports and Roboport Power Field functionality back with a Startup setting
---------------------------------------------------------------------------------------------------
Version: 1.3.10
Date: 2021.06.16
  Changes:
    - Added incompatible flag with BetterBots, in case it is ever updated
---------------------------------------------------------------------------------------------------
Version: 1.3.9
Date: 2021.06.14
  Changes:
    - Removed Wired Roboports and Roboport Power Field functionality
---------------------------------------------------------------------------------------------------
Version: 1.3.8
Date: 2020.12.06
  Features:
    - Updated to 1.1.x
---------------------------------------------------------------------------------------------------
Version: 1.3.7
Date: 2020.04.22
  Changes:
    - Updated the Reset Roboport procedure to destroy any remaining dummy entity on the map.
      This should help fix a very rare occasion of invisible power poles remaining on the
      map after destroying roboports either manually or with bots.
---------------------------------------------------------------------------------------------------
Version: 1.3.6
Date: 2020.04.09
  Bugfixes:
    - Added compatibility with Factorio version 0.18.18
---------------------------------------------------------------------------------------------------
Version: 1.3.5
Date: 2020.03.03
  Major Features:
    - Added support for Factorio 0.18
---------------------------------------------------------------------------------------------------
Version: 1.3.4
Date: 2019.12.09
  Bugfixes:
    - The mod is now compatible with Industrial Revolution mod
---------------------------------------------------------------------------------------------------
Version: 1.3.3
Date: 2019.11.27
  Bugfixes:
    - Improved compatibility with mods that modify the roboport prototype during data stage
---------------------------------------------------------------------------------------------------
Version: 1.3.2
Date: 2019.06.05
  Minor Features:
    - Added new icons for each level of the Advanced Roboports technology, to distinguish
      them more easily when filtering a Deconstruction Planner
  Bugfixes:
    - Fixed a bug with blueprints containing only tiles.
    - Fixed a bug with placing a blueprint from a blueprint book.
---------------------------------------------------------------------------------------------------
Version: 1.3.1
Date: 2019.06.04
  Major Features:
    - The Pipette Tool, Deconstruction Planner and Blueprint Tool now work as intended
    - Building roboports from blueprints now correctly retains all planned circuit network
      connections.
  Balancing:
    - The Worker Robot Power Cells technology now requires Utility Science Packs for tech
      level 3 and 4 to conform with other robot related technologies.
    - The Advanced Roboports 2 technology now requires Utility Science Packs instead
      of Production Science Packs.
    - The Roboport Sensor Array 1 technology now requires Utility Science Packs instead of
      Production Science Packs.
  Bugfixes:
    - The roboport upgrade procedure now correcly keeps the inventory exactly as it was
      before instead of simply adding the same number of items
---------------------------------------------------------------------------------------------------
Version: 1.3.0
Date: 2019.05.24
  Changes:
    - Added Factorio 0.17 support
  Bugfixes:
    - Reduced the performance hit of the charting procedure on maps with very high number of
      roboports. This should fix the occasional stutter that some players were experiencing
      exactly every 5 seconds
---------------------------------------------------------------------------------------------------
Version: 1.2.0
Date: 2018.03.08
  Changes:
    - Added a new Technology: "Roboport power field". Allows roboports to power nearby structures.
  Bugfixes:
    - Fixed an issue occuring when upgrading roboports connected to entities with more than one
      circuit connection (like combinators)
---------------------------------------------------------------------------------------------------
Version: 1.1.0
Date: 26. 02. 2018
  Changes:
    - Added a new Technology: "Wired Roboports". Allows roboports to be wired directly to nearby
      power poles and roboports.
    - Added a new Technology: "Roboport sensor array". Allows roboports to scan and reveal nearby
      sectors on the minimap.
    - Added a scheduler to any upgrade or reset operation on roboports to prevent the game from
      freezing in maps with a very high number of roboports. This scheduler comes with its own
      interface for monitoring the process of queued tasks. There are also two settings in the
      "Mod settings" section to enable or disable the GUI and to change the number of tasks to be
      processed per in-game tick.
---------------------------------------------------------------------------------------------------
Version: 1.0.3
Date: 2018.02.23
  Bugfixes:
    - Improved compatibility with other mods that create new entities based on roboports
---------------------------------------------------------------------------------------------------
Version: 1.0.2
Date: 2018.02.22
  Bugfixes:
    - Fixed the wrong signaling when upgrading or resetting roboports
---------------------------------------------------------------------------------------------------
Version: 1.0.1
Date: 2018.02.22
  Bugfixes:
    - Researching the roboport charge pads technology won't disconnect them from the circuit
      network anymore
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Date: 2018.02.21
  Major Features:
    - Initial release


settings.lua
data:extend{
    -- Startup Settings (loaded before the prototype stage)
    {
        type = "bool-setting",
        name = "betterbots-enable-wired-roboports",
        setting_type = "startup",
        default_value = false,
        order = "a-a"
    },

    --[[
    {
        type = "bool-setting",
        name = "betterbots-enable-charting-roboports",
        setting_type = "startup",
        default_value = false,
        --order = "
    },
    --]]

    -- Global settings (loaded before the control stage)
    {
        type = "bool-setting",
        name = "betterbots-show-gui",
        setting_type = "runtime-global",
        default_value = true,
        order = "a-a"
    },
    {
        type = "bool-setting",
        name = "betterbots-show-scheduler-gui",
        setting_type = "runtime-global",
        default_value = true,
        order = "b-a"
    },
    {
        type = "int-setting",
        name = "betterbots-scheduler-tasks-per-tick",
        setting_type = "runtime-global",
        default_value = 3,
        allowed_values = {1, 2, 3, 4, 5},
        order = "b-b"
    }
}


scheduler.lua
local mod_gui = require 'mod-gui'
require 'stdlib/event/event'

Scheduler = { }

local show_scheduler_gui = settings.global["betterbots-show-scheduler-gui"].value
local n_tpt = settings.global["betterbots-scheduler-tasks-per-tick"].value

Event.register(
    {defines.events.on_runtime_mod_setting_changed},
    function(event)
        local player = game.players[event.player_index]

        show_scheduler_gui = settings.global["betterbots-show-scheduler-gui"].value
        n_tpt = settings.global["betterbots-scheduler-tasks-per-tick"].value

        if not show_scheduler_gui then
            local main_frame = mod_gui.get_frame_flow(player).scheduler_summary_frame
            if main_frame and main_frame.valid then main_frame.destroy() end
        end
    end
)

function Scheduler._draw_gui()
    if show_scheduler_gui then
        for i, player in pairs(game.connected_players) do
            local main_frame = mod_gui.get_frame_flow(player).scheduler_summary_frame

            if main_frame then
                main_frame.destroy()
            end

            if not global.scheduler_queues or (#global.scheduler_queues == 0) then return end

            main_frame = mod_gui.get_frame_flow(player).add({type = "frame", name = "scheduler_summary_frame", direction = "vertical"})
            main_frame.add({type = "label", name = "scheduler_summary_label", caption = "Scheduled Actions", style = "bold_label"})

            for i, queue in pairs(global.scheduler_queues) do
                local queue_flow = main_frame.add({type = "flow", name = "scheduler_queue_flow_"..i, direction = "horizontal"})
                local style = "label"
                local caption

                if type(queue.label) == "string" then
                    caption = {"", queue.label}
                else
                    caption = table.deepcopy(queue.label)
                end

                table.insert(caption, 2, i..") ")

                if i == 1 then
                    style = "caption_label"
                end

                queue_flow.add({type = "label", name = "scheduler_summary_label_"..i, caption = caption, style = style})

                if (#queue.tasks > 5) then
                    queue_flow.add({type = "label", name = "scheduler_summary_n_tasks_"..i, caption = #queue.tasks})
                end
            end
        end
    end
end

function Scheduler._remove_queue(position)
    position = position or 1

    return table.remove(global.scheduler_queues, position)
end

function Scheduler._tick(event)
    Scheduler._draw_gui()

    if not global.scheduler_queues or (#global.scheduler_queues == 0) then return end

    if not global.scheduler_queues[1].tasks or (#global.scheduler_queues[1].tasks == 0) then
        Scheduler._remove_queue()
        return
    end

    local n_tasks = #global.scheduler_queues[1].tasks
    local tpt = n_tpt or 3

    for i = 1, tpt, 1 do
        local t = table.remove(global.scheduler_queues[1].tasks, 1)
        if not t then break end
        local a

        if t.args then
            if type(t.args) == "table" then
                t.func(unpack(t.args))
            else
                t.func(t.args)
            end
        end
    end
end

function Scheduler.queue_task(f, p, in_queue)
    if not global.scheduler_queues then
        global.scheduler_queues = { }
    end

    local queue = 1

    if not in_queue then
        if not Scheduler._open_queue then
            Scheduler.open_queue('Other tasks')
        end

        queue = Scheduler._open_queue
    end

    table.insert(global.scheduler_queues[queue].tasks, {func = f, args = p})
end

function Scheduler.open_queue(label)
    if not global.scheduler_queues then
        global.scheduler_queues = { }
    end

    if Scheduler._open_queue then Scheduler.close_queue() end

    table.insert(global.scheduler_queues, {label = label, tasks = { }})
    Scheduler._open_queue = #global.scheduler_queues
end

function Scheduler.close_queue()
    if not global.scheduler_queues or not Scheduler._open_queue then return end

    Scheduler._open_queue = nil
end

Event.register(
    {defines.events.on_tick},
    Scheduler._tick
)


migrations.lua
--[[
This is a dirty way to track migrations, because the global table is not available during the migration
scripts stage, so I can't access the entity data if I have to...

From the API:
on_configuration_changed:
Register a function to be run when mod configuration changes. This is called any time the game
version changes, prototypes change, startup mod settings change, and any time mod versions change
including adding or removing mods.
--]]
Event.register(
    {Event.core_events.init, Event.core_events.configuration_changed},
    function(event)
        if not global._migrations then
            global._migrations = { }
        end

        -- 1.0.3 to 1.1.0
        if global._migrations and not global._migrations.migrated_dummy_bps then
            game.print("BetterBots - migrating 1.0.3 to 1.1.0")
            global._migrations.migrated_dummy_bps = true

            for _, surface in pairs(game.surfaces) do
                local ports = surface.find_entities_filtered({type = 'roboport'})

                for _, port in pairs(ports) do
                    if port.name:starts_with("betterbots-roboport") then
                        local port_data = Entity.get_data(port)

                        if port_data and port_data.blueprint_roboport then
                            local bp_port = port_data.blueprint_roboport

                            if not port_data.dummy_entities then port_data.dummy_entities = { } end
                            if port_data.dummy_entities.blueprint_roboport and port_data.dummy_entities.blueprint_roboport.valid and not Entity._are_equal(port_data.blueprint_roboport, port_data.dummy_entities.blueprint_roboport) then
                                port_data.dummy_entities.blueprint_roboport.destroy()
                            end

                            port_data.dummy_entities.blueprint_roboport = bp_port
                            port_data.blueprint_roboport = nil

                            Entity.set_data(port, port_data)
                        end
                    end
                end
            end
        end

        -- 1.2.0 to 1.3.0
        if global._migrations and not global._migrations.migrated_charting_code then
            game.print("BetterBots - migrating 1.2.0 to 1.3.0")
            global._migrations.migrated_charting_code = true

            global.charting_ports = { }

            table.each(game.surfaces,
                function(surface)
                    local ports = surface.find_entities_filtered({type = 'roboport'})

                    table.each(ports,
                        function(port)
                            if not port.valid then return end

                            if port.name:starts_with("betterbots-roboport") then
                                local port_data = Entity.get_data(port) or { }

                                if port_data and port_data.charting_range and port_data.charting_range > 0 then
                                    table.insert(global.charting_ports, port)
                                end
                            end
                        end
                    )
                end
            )
        end

        -- 1.3.0 to 1.3.1
        if global._migrations and not global._migrations.migrated_dummy_blueprint_roboport then
            game.print("Betterbots - migrating 1.3.0 to 1.3.1 - Dummy blueprintable entities destroyed")

            global._migrations.migrated_dummy_blueprint_roboport = true
        end
    end
)


prototypes/entity/roboport.lua
require("circuit-connector-sprites")

local roboports = { }

local pads_offsets = {
    {{-1.5, -0.5}, {1.5, -0.5}, {1.5, 1.5}, {-1.5, 1.5}, {0, 2}},
    {{-1.5, -0.5}, {1.5, -0.5}, {1.5, 1.5}, {-1.5, 1.5}, {0, 2}, {0, -1}},
    {{-1.5, -0.5}, {1.5, -0.5}, {1.5, 1.5}, {-1.5, 1.5}, {0, 2}, {0, -1}, {-2, 0.5}},
    {{-1.5, -0.5}, {1.5, -0.5}, {1.5, 1.5}, {-1.5, 1.5}, {0, 2}, {0, -1}, {-2, 0.5}, {2, 0.5}}
}

for robopads_tech = 1, 4 do
    local port = table.deepcopy(data.raw['roboport']['roboport'])

    local port_name = 'betterbots-roboport_' .. robopads_tech
    port.charging_offsets = pads_offsets[robopads_tech]
    port.charging_energy = 1000 + 200 * robopads_tech .. "kW"

    local buffer_rate = 1000 + (1000 + 200 * robopads_tech) * (robopads_tech + 4)
    port.energy_source.input_flow_limit = buffer_rate .. "kW"

    --[[

    Buffer_capacity is calculated to give 25 seconds of full energy flow to bots when disconnected from
    the power line, like vanilla roboports, but doing so results in a huge power spike when upgrading all the roboports
    together, especially on the last tier of research. I preferred to keep it disabled for this reason. Anyway, it
    should only be noticeable when there's a shortage of electricity, since the max possible drain from all charge pads
    is always less than the max energy input.

    port.energy_source.buffer_capacity = 25 * (buffer_rate - 1000) .. "kJ"

    --]]

    local energy_use = 50 * robopads_tech -- + (50 * robopads_tech * adv_robo_tech * 3)
    port.localised_name = {"entity-name.roboport"}

    port.name = port_name -- .. '_' .. adv_robo_tech
    port.energy_usage = energy_use .. "kW"
    port.placeable_by = {item = "roboport", count = 1}
    port.minable = {mining_time = 0.1, result = "roboport", count = 1}
    port.icon = "__BetterBotsWLK__/graphics/icons/roboport_" .. robopads_tech .. ".png"
    port.icon_size = 32

    table.insert(roboports, port)
end

if (settings.startup["betterbots-enable-wired-roboports"].value) then
    --[[

    Dummy Power Pole entity

    I wish we could wire any entity with copper cables instead of being forced into this mess....

    --]]

    local dummy_pole = {
        type = "electric-pole",
        name = "betterbots-dummy-pole",
        icon = "__base__/graphics/icons/big-electric-pole.png",
        icon_size = 32,
        drawing_box = {{0,0}, {0,0}},
        maximum_wire_distance = 30,
        supply_area_distance = 0,
        pictures = {
            filename = "__core__/graphics/empty.png",
            priority = "low",
            width = 1,
            height = 1,
            direction_count = 1,
            apply_projection = false,
            shift = {0.0, 0.0}
        },
        connection_points = {
            {
                shadow = {
                    copper = {-0.5, 0.5}
                },
                wire = {
                    copper = {-0.5, 0.5}
                }
            }
        },
        order = "c-a",
        radius_visualisation_picture =
        {
            filename = "__core__/graphics/empty.png",
            width = 1,
            height = 1,
            priority = "low"
        },
    }

    table.insert(roboports, dummy_pole)

    local powering_pole = {
        type = "electric-pole",
        name = "betterbots-powering-pole",
        icon = "__base__/graphics/icons/big-electric-pole.png",
        icon_size = 32,
        drawing_box = {{0,0}, {0,0}},
        maximum_wire_distance = 30,
        supply_area_distance = 4,
        pictures = {
            filename = "__core__/graphics/empty.png",
            priority = "low",
            width = 1,
            height = 1,
            direction_count = 1,
            apply_projection = false,
            shift = {0.0, 0.0}
        },
        connection_points = {
            {
                shadow = {
                    copper = {-0.5, 0.5}
                },
                wire = {
                    copper = {-0.5, 0.5}
                }
            }
        },
        order = "c-a",
        radius_visualisation_picture =
        {
            filename = "__base__/graphics/entity/small-electric-pole/electric-pole-radius-visualization.png",
            width = 12,
            height = 12,
            priority = "extra-high-no-scale"
        },
    }

    table.insert(roboports, powering_pole)
end

data:extend(roboports)


prototypes/technology.lua
-- ROBOPORT RELATED TECHNOLOGIES
data:extend({
    -- Advanced roboports
    {
        type = "technology",
        name = "roboport-charge-pads-1",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/roboport-charge-pads.png",
        effects = { },
        prerequisites = {
            "logistic-robotics",
            "construction-robotics"
        },
        unit = {
            count = 100,
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1},
            },
            time = 30
        },
        upgrade = true,
        order = "c-k-f-a",
    },
    {
        type = "technology",
        name = "roboport-charge-pads-2",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/roboport-charge-pads.png",
        effects = { },
        prerequisites = {
            "roboport-charge-pads-1"
        },
        unit = {
            count = 150,
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1},
                {"utility-science-pack", 1},
            },
            time = 30
        },
        upgrade = true,
        order = "c-k-f-b",
    },
    {
        type = "technology",
        name = "roboport-charge-pads-3",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/roboport-charge-pads.png",
        effects = { },
        prerequisites = {
            "roboport-charge-pads-2"
        },
        unit = {
            count = 250,
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1},
                {"production-science-pack", 1},
                {"utility-science-pack", 1},
            },
            time = 60
        },
        upgrade = true,
        order = "c-k-f-c",
    },
    {
        type = "technology",
        name = "roboport-charge-pads-4",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/roboport-charge-pads.png",
        effects = { },
        prerequisites = {
            "roboport-charge-pads-3"
        },
        unit = {
            count = 350,
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1},
                {"production-science-pack", 1},
                {"utility-science-pack", 1},
                {"space-science-pack", 1}
            },
            time = 60
        },
        upgrade = true,
        order = "c-k-f-d",
    },

    -- Roboport Sensor Array
    {
        type = "technology",
        name = "charting-roboports-1",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/charting-roboports.png",
        effects = { },
        prerequisites = {
            "roboport-charge-pads-1",
            --"electric-energy-distribution-2",
            "modules"
        },
        unit = {
            count = 200,
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1},
                {"production-science-pack", 1},
            },
            time = 30
        },
        upgrade = true,
        order = "c-k-f-g",
    },
    {
        type = "technology",
        name = "charting-roboports-2",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/charting-roboports.png",
        effects = { },
        prerequisites = {
            "charting-roboports-1",
            "roboport-charge-pads-4"
        },
        unit = {
            count = 500,
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1},
                {"production-science-pack", 1},
                {"utility-science-pack", 1},
                {"space-science-pack", 1}
            },
            time = 30
        },
        upgrade = true,
        order = "c-k-f-h",
    }
})

if (settings.startup["betterbots-enable-wired-roboports"].value) then
    data:extend({
        -- Wired Roboports
        {
            type = "technology",
            name = "wired-roboports",
            icon_size = 128,
            icon = "__BetterBotsWLK__/graphics/technology/wired-roboports.png",
            effects = { },
            prerequisites = {
                "roboport-charge-pads-2",
                "electric-energy-distribution-2",
                "modules"
            },
            unit = {
                count = 250,
                ingredients = {
                    {"automation-science-pack", 1},
                    {"logistic-science-pack", 1},
                    {"chemical-science-pack", 1},
                    {"production-science-pack", 1},
                    {"utility-science-pack", 1},
                },
                time = 30
            },
            --upgrade = true,
            order = "c-k-f-e",
        },
        {
            type = "technology",
            name = "roboport-power-field",
            icon_size = 128,
            icon = "__BetterBotsWLK__/graphics/technology/wired-roboports.png",
            effects = { },
            prerequisites = {
                "wired-roboports",
            },
            unit = {
                count = 450,
                ingredients = {
                    {"automation-science-pack", 1},
                    {"logistic-science-pack", 1},
                    {"chemical-science-pack", 1},
                    {"production-science-pack", 1},
                    {"utility-science-pack", 1},
                    {"space-science-pack", 1}
                },
                time = 30
            },
            --upgrade = true,
            order = "c-k-f-f",
        }
    })
end

-- BOTS RELATED TECHNOLOGIES
data:extend({
    -- Worker Robot Power Cells
    {
        type = "technology",
        name = "worker-robots-battery-1",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/worker-robots-battery.png",
        effects = {
            {
                type = "worker-robot-battery",
                modifier = 0.15
            }
        },
        prerequisites = {
            "robotics"
        },
        unit = {
            count = 50,
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1}
            },
            time = 30
        },
        upgrade = true,
        order = "c-k-f-a",
    },
    {
        type = "technology",
        name = "worker-robots-battery-2",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/worker-robots-battery.png",
        effects = {
            {
                type = "worker-robot-battery",
                modifier = 0.2
            }
        },
        prerequisites = {
            "worker-robots-battery-1"
        },
        unit = {
            count = 100,
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1}
            },
            time = 30
        },
        upgrade = true,
        order = "c-k-f-b"
    },
    {
        type = "technology",
        name = "worker-robots-battery-3",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/worker-robots-battery.png",
        effects = {
            {
                type = "worker-robot-battery",
                modifier = 0.25
            }
        },
        prerequisites = {
            "worker-robots-battery-2"
        },
        unit = {
            count = 150,
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1},
                {"utility-science-pack", 1}
            },
            time = 60
        },
        upgrade = true,
        order = "c-k-f-c"
    },
    {
        type = "technology",
        name = "worker-robots-battery-4",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/worker-robots-battery.png",
        effects = {
            {
                type = "worker-robot-battery",
                modifier = 0.35
            }
        },
        prerequisites = {
            "worker-robots-battery-3"
        },
        unit = {
            count = 250,
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1},
                {"utility-science-pack", 1}
            },
            time = 60
        },
        upgrade = true,
        order = "c-k-f-d"
    },
    {
        type = "technology",
        name = "worker-robots-battery-5",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/worker-robots-battery.png",
        effects = {
            {
                type = "worker-robot-battery",
                modifier = 0.35
            }
        },
        prerequisites = {
            "worker-robots-battery-4"
        },
        unit = {
            count = 500,
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1},
                {"production-science-pack", 1},
                {"utility-science-pack", 1}
            },
            time = 60
        },
        upgrade = true,
        order = "c-k-f-e"
    },
    {
        type = "technology",
        name = "worker-robots-battery-6",
        icon_size = 128,
        icon = "__BetterBotsWLK__/graphics/technology/worker-robots-battery.png",
        effects = {
            {
                type = "worker-robot-battery",
                modifier = 0.45
            }
        },
        prerequisites = {
            "worker-robots-battery-5"
        },
        unit = {
            count_formula = "2^(L-6)*1000",
            ingredients = {
                {"automation-science-pack", 1},
                {"logistic-science-pack", 1},
                {"chemical-science-pack", 1},
                {"production-science-pack", 1},
                {"utility-science-pack", 1},
                {"space-science-pack", 1}
            },
            time = 60
        },
        max_level = "infinite",
        upgrade = true,
        order = "c-k-f-e"
    },
})
